### ISSUE TITLE
The random_state of the final model is different than the during search

### ISSUE THREAD

**Issue by JoshHeitzman** [EpistasisLab/tpot#894]
https://github.com/EpistasisLab/tpot/issues/893 provides further context as this was found during the investigation of a specific cause of an exception being thrown during the final fit of the top pipeline after the search is complete.

In this case the the pipelines were not precisely the same during search and during the final fit, with the random_state being None in the final fit, but 42 during the search.  The TPOT classifer was configured with a random_state of 1, so that appears to have been ignored when configuring the pipeline during both the search and the final fit.

This is the pipeline during the search:

Pipeline(memory=None,
         steps=[('stackingestimator',
                 StackingEstimator(estimator=LogisticRegression(C=0.01,
                                                                class_weight=None,
                                                                dual=True,
                                                                fit_intercept=True,
                                                                intercept_scaling=1,
                                                                l1_ratio=None,
                                                                max_iter=100,
                                                                multi_class='warn',
                                                                n_jobs=None,
                                                                penalty='l2',
                                                                random_state=42,
                                                                solver='warn',
                                                                tol=0.0001,
                                                                verbose=0,
                                                                warm_start=False))),
                ('bernoullinb',
                 BernoulliNB(alpha=0.001, binarize=0.0, class_prior=None,
                             fit_prior=False))],
         verbose=False)

and this if the pipeline for the final fit:

Pipeline(memory=None,
         steps=[('stackingestimator',
                 StackingEstimator(estimator=LogisticRegression(C=0.01,
                                                                class_weight=None,
                                                                dual=True,
                                                                fit_intercept=True,
                                                                intercept_scaling=1,
                                                                l1_ratio=None,
                                                                max_iter=100,
                                                                multi_class='warn',
                                                                n_jobs=None,
                                                                penalty='l2',
                                                                random_state=None,
                                                                solver='warn',
                                                                tol=0.0001,
                                                                verbose=0,
                                                                warm_start=False))),
                ('bernoullinb',
                 BernoulliNB(alpha=0.001, binarize=0.0, class_prior=None,
                             fit_prior=False))],
         verbose=False)

During final fit the StackingEstimator transform introduces NaNs/infinites, but during the search I had modified to code, as mentioned in 893, to fit the pipeline to the full data set as is done in final fit.  The transform the StackingEstimator did during full fit I had hacked into search did not result in the check for NaNs/infinites raising.

The only difference I can see between the pipelines is value being provided to the random_state.

**Comment by JoshHeitzman** [comment]
The difference in the random_state of the pipelines for search and the final fit results TPOTBase._preprocess_individuals having the following block of code after calling self._toolbox.compile while _summary_of_best_pipeline lacks an equivalent:

                    # Fix random state when the operator allows
                    self._set_param_recursive(sklearn_pipeline.steps, 'random_state', 42)
                    # Setting the seed is needed for XGBoost support because XGBoost currently stores
                    # both a seed and random_state, and they're not synced correctly.
                    # XGBoost will raise an exception if random_state != seed.
                    if 'XGB' in sklearn_pipeline_str:
                        self._set_param_recursive(sklearn_pipeline.steps, 'seed', 42)

After adding a equivalent block in _summary_of_best_pipeline I no longer see an exception raised during the final fit of that model.

**Comment by weixuanfu** [comment]
We intentionally set the random_state of 42 during search and there should no random_state in exported pipeline. And the random_state in TPOTBase is for randomly generating pipeline or crossover/mutation. We encountered a few similar issues before, maybe we should use the `random_state` in TPOTBase for search.

**Comment by JoshHeitzman** [comment]
@weixuanfu I would expect using the random_state supplied to TPOTBase for the pipelines during search would allow users to get broader coverage in the searches when they are interested in trying different random seeds.

I'm not sure why the final pipeline (in memory or exported) would want to use None for the random_state.  I haven't checked for reproducibility yet on the project I'm currently working on, but on one for a previous employer I was seeing some cases of the same data and the same configuration for TPOT producing pipelines that would make different predictions, so I'm now wondering if random_state being None was contributing to that (don't have access to verify though).

**Comment by weixuanfu** [comment]
I think it should related to random_state=None in final pipeline. 

We have some discussions about it on this related issue #514.

**Comment by jhmenke** [comment]
Does that mean i can recreate the TPOT model exactly by using the same CV and a random_state=42 (wherever possible)?

**Comment by weixuanfu** [comment]
@jhmenke TPOT should regenerate the same best model with the same arguments in the same running environment.

**Comment by weixuanfu** [comment]
I think this issue should be fixed in PR #945

**Comment by weixuanfu** [comment]
The issue has been fixed in a new version of TPOT (v0.11.0). Please feel free to reopen it if you have any questions or suggestions.
